src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/EditorAsyncCompletionItemManager.cs

```diff
@@ -0,0 +1,614 @@
+            var needToFilter = selectedFilters.Length > 0 && selectedFilters.Length < data.SelectedFilters.Length;
+
+            var initialListOfItemsToBeIncluded = new List<FilterResult>();
+            foreach (var item in data.InitialSortedList)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                if (needToFilter && ShouldBeFilteredOutOfCompletionList(item, selectedFilters))
+                {
+                    continue;
+                }
+
+                if (MatchesFilterText(item.FilterText, item.DisplayText, GetMatchPriority(item), filterText, data.Trigger.Reason))
+                {
+                    initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: true));
+                }
+                else
+                {
+                    if (data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Deletion ||
+                        data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Backspace ||
+                        data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Invoke ||
+                        filterText.Length <= 1)
+                    {
+                        initialListOfItemsToBeIncluded.Add(new FilterResult(item, filterText, matchedFilterText: false));
+                    }
+                }
+            }
+
+            // If the session was created/maintained out of Roslyn, e.g. debugger.; no properties are set and we should use data.Snapshot.
+            var snapshotForDocument = data.InitialSortedList
+                .FirstOrDefault(i => i.Properties.ContainsProperty(CompletionItemSource.TriggerBuffer))?.
+                Properties.GetProperty<ITextBuffer>(CompletionItemSource.TriggerBuffer)?.CurrentSnapshot ?? data.Snapshot;
+
+            var document = snapshotForDocument.GetOpenDocumentInCurrentContextWithChanges();
+
+            if (document == null)
+            {
+                var listWithSelections = GetHighlightedList(initialListOfItemsToBeIncluded, filterText);
+                return new EditorCompletion.FilteredCompletionModel(listWithSelections.ToImmutableArray(), 0, data.SelectedFilters, EditorCompletion.UpdateSelectionHint.NoChange, centerSelection: true, uniqueItem: default);
+            }
+
+            if (initialListOfItemsToBeIncluded.Count == 0)
+            {
+                var completionService = document.GetLanguageService<CompletionService>();
+                var completionRules = completionService.GetRules();
+                return HandleAllItemsFilteredOut(data.Trigger.Reason, data.SelectedFilters, selectedFilters, completionRules);
+            }
+
+            // If this was deletion, then we control the entire behavior of deletion ourselves.
+            if (data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Deletion || data.Trigger.Reason == EditorCompletion.CompletionTriggerReason.Backspace)
```

Comment:
is there a reason we're not mapping back to the Roslyn concepts and then keeping the logic the same as before? ideally even using the exact same functions?

Reply:
In this method, we need only the reason from the trigger. I think it is simpler to support the editor reason than make a conversion and then support our own.

---
In reply to: [222086529](https://github.com/dotnet/roslyn/pull/29016#discussion_r222086529) [](ancestors = 222086529)

Reply:
it makes thsi so much harder to review though.... :-/